workflow/rules/dba.smk

```diff
@@ -106,218 +84,206 @@ RankColIDR = {
+            os.mkdir(params.fldr, mode=0o775)
+
+        json_construct = dict()
+        json_construct['queries'] = []
+        json_construct['show_attributes'] = ["gene_id", "gene_name", "gene_type"]
+        json_construct["priority"] = "Yes"
+        json_construct['gtf'] = gtf
+        json_construct['bed'] = input[0]
+
+        base_query = {
+            "feature": "gene",
+            "filter.attribute": "gene_type",
+            "attribute.value": "protein_coding", 
+            "feature.anchor": "start" 
+        }
+
+        if assay == 'cfchip':
+            if wildcards.type == 'protTSS':
+                for _d in (3000, 10000, 100000):
+                    this_q = base_query.copy()
+                    this_q['distance'] = _d
+                    json_construct['queries'].append(this_q)
+        else:
+            if wildcards.type == 'prot':
+                for _d in (5000, 100000):
+                    this_q = base_query.copy()
+                    del this_q["feature.anchor"]
+                    this_q['distance'] = _d
+                    json_construct['queries'].append(this_q)
+            elif wildcards.type == 'genes':
+                this_query = {}
+                this_query['feature'] = 'gene'
+                for _d in (5000, 100000):
+                    this_q = base_query.copy()
+                    del this_q["feature.anchor"]
+                    del this_q["filter.attribute"]
+                    del this_q["attribute.value"]
+                    this_q['distance'] = _d
+                    json_construct['queries'].append(this_q)
+            elif wildcards.type == 'protSEC':
+                # distance, feature.anchor
+                query_values = (
+                    ([3000, 1000], "start"), 
+                    (3000,         "end"), 
+                    (100000,       "center"), 
+                    (100000,       None)
+                )
+                for _distance, feature_anchor in query_values:
+                    this_q = base_query.copy()
+                    del this_q["feature.anchor"]
+                    if feature_anchor: 
+                        this_q["feature.anchor"] = feature_anchor
+                    this_q['distance'] = _distance
+                    json_construct['queries'].append(this_q)
+            elif wildcards.type == 'protTSS':
+                for _d in ([3000, 1000], 10000, 100000):
+                    this_q = base_query.copy()
+                    this_q['distance'] = _d
+                    json_construct['queries'].append(this_q)
+
+        with open(output.json, 'w') as jo:
+            json.dump(json_construct, jo, indent=4)
+            jo.close()
+
+        if not os.path.exists(output.json):
+            raise FileNotFoundError(output.json + " does not exist!")
+        shell.prefix(f"module load {uropaver};")
+        shell("uropa -i " + output.json + " -p " + params.outroot + " -t " + str(threads) + " -s")
```

Comment:
I think the run statements are a great to bundle python preprocessing with the targeted software execution and it has been supported in snakemake for a very long time, if the container doesn't have python setup properly or isn't initialized as a login shell with path directives you will probably have trouble with the interpreter resolution, but I would argue that is a problem with the container or its orchestration moreso than any bug with the snakemake `run` directive. shell and shell.prefix are all from the snakemake namespace, this particular rule is vanilla python3 with snakemake functions (an enforced requirement).

Reply:
Yeah, in the past we have run into issues with using the `run` and `container` directive together. Here is a **Note** from the docs about it:
![image](https://github.com/user-attachments/assets/c4127967-11b3-47ec-b359-95ee56cacdb9)

The long-term goal is to use containers at each step and remove the module dependencies. With that being said, we don't need to fix this right now but in the future, we may need to just move some of this logic into a script. In the past, we have also used EOF statements to create a python script on the fly. We could use that instead to create the script that creates the config file and then run uropa using the `shell` directive.

Reply:
Here is some information from the latest 7.X version, since the `singularity` directive was eventually replaced by the `container` directive a while ago:
https://snakemake.readthedocs.io/en/v7.32.3/snakefiles/deployment.html#running-jobs-in-containers

Reply:
Yeah so run has to execute in the python interpreter that the global namespace snakemake is in. Given that this rule still uses env-modules I think it is written the best way given the dynamic creation of a query JSON. When uropa gets containerized, yeah, then moving to `shell` or `script` makes sense.

Reply:
Or split this rule into two, one for json creation with `run` and NO container, and one for uropa with `shell`.

Reply:
Yeah, I would just split the rule or just copy the json creation logic into a script that can take command-line arguments for input/output.

Reply:
I am going to resolve this for now, we can file an issue to split the rule after merging.